File: DSA/JAVA/Queue/Circular-Queue.cpp
#include <iostream>
using namespace std;

class CircularQueue {
    int size;
    int* q;
    int front;
    int rear;

public:
    CircularQueue(int size) {
        this->size = size;
        q = new int[size];
        front = -1;
        rear = -1;
    }

    // Insert element at rear
    void enQueue(int data) {
        if ((rear + 1) % size == front) {
            cout << "Overflow" << endl;
            return;
        }

        // Empty queue
        if (isEmpty()) {
            front = rear = 0;
            q[rear] = data;
        } else {
            rear = (rear + 1) % size;
            q[rear] = data;
        }
    }

    // Delete element from front
    void deQueue() {
        if (front == -1) {
            cout << "Underflow" << endl;
            return;
        }

        if (front == rear) {
            front = rear = -1;
        } else {
            front = (front + 1) % size;
        }
    }

    // Check if queue is empty
    bool isEmpty() {
        return (rear == -1);
    }

    // Get the front element
    int getFront() {
        if (front == -1) {
            return -1;
        }
        return q[front];
    }

    // Traverse and print the queue
    void traverse() {
        if (isEmpty()) {
            cout << "Empty Queue" << endl;
            return;
        }

        int i = front;
        while (true) {
            cout << q[i] << " ";
            if (i == rear) break;
            i = (i + 1) % size;
        }
        cout << endl;
    }

    ~CircularQueue() {
        delete[] q;
    }
};

int main() {
    CircularQueue queue(5);
    queue.enQueue(10);
    queue.enQueue(20);
    queue.enQueue(30);

    queue.traverse();
    queue.deQueue();
    queue.traverse();
    cout << boolalpha << queue.isEmpty() << endl;
    cout << queue.getFront() << endl;

    return 0;
}

File: DSA/JAVA/Queue/Circular-Queue.java

class CircularQueue{
    int size;
    int [] q;
    int front;
    int rare;

    public CircularQueue(int size){
        this.size = size;
        q = new int[size];
        front = -1;
        rare = -1;
    }

//    to insert the element at the rare end
    public void enQueue(int data){
        if (rare + 1 == front){
            System.out.println("Overflow");
            return;
        }
//        Empty Queue
        if (isEmpty()){
            front++;
            rare++;
            q[rare] = data;

        }else{
            rare = (rare + 1) % size;
            q[rare] = data;
        }
    }

    // to delete the element at front end
    public void deQueue(){
        if (front == -1){
            System.out.println("Underflow");
            return;
        }

        if (front == rare) { // single element in the queue
            front = rare = -1;
        }else{
            front = (front + 1) % size;
        }
    }

    // checks if queue is empty or not
    public boolean isEmpty(){
        return (rare == -1);
    }

    // first element to be deleted
    public int front(){
        if (front == -1){
            return -1;
        }

        return q[front];
    }

    public void traverse(){

        if (isEmpty()){
            System.out.println("Empty Queue");
            return;
        }

        for (int i = front; i != rare; i = (i + 1) % size){
            System.out.print(q[i] + " ");
        }
        System.out.println();
    }

}


public class Main {
    public static void main(String[] args) {

        CircularQueue queue = new CircularQueue(5);
        queue.enQueue(10);
        queue.enQueue(20);
        queue.enQueue(30);

        queue.traverse();
        queue.deQueue();
        queue.traverse();
        System.out.println(queue.isEmpty());
        System.out.println(queue.front());
    }
}